cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/resource/volume/AwsVolumeResourceBuilder.java

```diff
@@ -367,21 +371,128 @@ public CloudResource delete(AwsContext context, AuthenticatedContext auth, Cloud
 
         boolean anyDeleted = cloudResourceStatuses.stream().map(CloudResourceStatus::getStatus).anyMatch(DELETED::equals);
         AmazonEc2Client client = getAmazonEC2Client(auth);
-        if (!volumeSetAttributes.getDeleteOnTermination() && !anyDeleted) {
-            LOGGER.debug("Volumes will be preserved.");
-            resource.setStatus(CommonStatus.DETACHED);
-            volumeSetAttributes.setDeleteOnTermination(Boolean.TRUE);
-            turnOffDeleteOnterminationOnAttachedVolumes(resource, cloudResourceStatuses, client);
-            resource.putParameter(CloudResource.ATTRIBUTES, volumeSetAttributes);
-            resourceNotifier.notifyUpdate(resource, auth.getCloudContext());
-            throw new PreserveResourceException("Resource will be preserved for later reattachment.");
+        if (!volumeSetAttributes.getDeleteOnTermination()) {
+            if (anyDeleted) {
+                LOGGER.debug(String.format("Volumes should be preserved for instance (%s), but some of them are already terminated. " +
+                        "Trying to recover/recreate them.", resource.getInstanceId()));
+                List<Volume> volumes = tryRecoverVolumeSet(context, auth, resource, volumeSetAttributes);
+                LOGGER.debug(String.format("Recovered and preserved volumes: %s", volumes));
+                if (volumes != null && !volumes.isEmpty()) {
+                    volumeSetAttributes.setVolumes(volumes);
+                }
+            }
+            setVolumeDetachParameters(auth, resource, volumeSetAttributes, cloudResourceStatuses, client);
         }
         deleteOrphanedVolumes(cloudResourceStatuses, client);
         turnOnDeleteOnterminationOnAttachedVolumes(resource, cloudResourceStatuses, client);
 
         return null;
     }
 
+    private void setVolumeDetachParameters(AuthenticatedContext auth, CloudResource resource, VolumeSetAttributes volumeSetAttributes,
+            List<CloudResourceStatus> cloudResourceStatuses, AmazonEc2Client client) throws PreserveResourceException {
+        LOGGER.debug("Volumes will be preserved.");
+        resource.setStatus(CommonStatus.DETACHED);
+        volumeSetAttributes.setDeleteOnTermination(Boolean.TRUE);
+        turnOffDeleteOnterminationOnAttachedVolumes(resource, cloudResourceStatuses, client);
+        resource.putParameter(CloudResource.ATTRIBUTES, volumeSetAttributes);
+        resourceNotifier.notifyUpdate(resource, auth.getCloudContext());
+        throw new PreserveResourceException("Resource will be preserved for later reattachment.");
```

Comment:
I might move this exception throwing outside of the setter to make the code more readable.